be/src/exec/pipeline/exchange/local_exchange.h

```diff
@@ -119,12 +122,16 @@ class PartitionExchanger final : public LocalExchanger {
                        LocalExchangeSourceOperatorFactory* source, const TPartitionType::type part_type,
                        const std::vector<ExprContext*>& _partition_expr_ctxs, size_t num_sinks);
 
+    Status prepare(RuntimeState* state) override;
+    void close(RuntimeState* state) override;
+
     Status accept(const vectorized::ChunkPtr& chunk, int32_t sink_driver_sequence) override;
 
 private:
     // Used for local shuffle exchanger.
     // The sink_driver_sequence-th local sink operator exclusively uses the sink_driver_sequence-th partitioner.
     // TODO(lzh): limit the size of _partitioners, because it will cost too much memory when dop is high.
+    std::vector<ExprContext*> _partition_exprs;
     std::vector<Partitioner> _partitioners;
```

Comment:
Prepare, open, close partition exprs in the `LocalShuffleSinkOperatorFactory`. 
The reason is that we forget to close some parttion exprs of local shuffle some times, because preparation logical is all over the place.

Reply:
It is safe, because 
1. Prepare, open and close of ExprContext is idempotent, they only perform real prepare, open and close operations once.
2. `LocalExchangeSinkOperator::close` is invoked when all the drivers of this fragment instances are finished.

Reply:
Yes, `Partitioner` just uses `partition_expr_ctxs`, not prepare, open, and close.